src/software/amazon/ion/impl/IonReaderTextSystemX.java

```diff
@@ -85,25 +126,100 @@ public IonSystem getSystem()
+            }
+
+            @Override
+            boolean isLong(String image, int len)
+            {
+                return valueWithinBounds(image, len, MIN_LONG_IMAGE, MAX_LONG_IMAGE);
+            }
+
+        },
+        HEX
+        {
+
+            @Override
+            boolean isInt(String image, int len)
+            {
+                return valueWithinBounds(image, len, MIN_HEX_INT_IMAGE, MAX_HEX_INT_IMAGE);
+            }
+
+            @Override
+            boolean isLong(String image, int len)
+            {
+                return valueWithinBounds(image, len, MIN_HEX_LONG_IMAGE, MAX_HEX_LONG_IMAGE);
+            }
+
+        },
+        BINARY
+        {
+
+            @Override
+            boolean isInt(String image, int len)
+            {
+                return valueWithinBounds(image, len, MIN_BINARY_INT_IMAGE, MAX_BINARY_INT_IMAGE);
+            }
+
+            @Override
+            boolean isLong(String image, int len)
+            {
+                return valueWithinBounds(image, len, MIN_BINARY_LONG_IMAGE, MAX_BINARY_LONG_IMAGE);
+            }
+
+        };
+
+        private static final char[] MAX_INT_IMAGE = Integer.toString(Integer.MAX_VALUE).toCharArray();
+        private static final char[] MIN_INT_IMAGE = Integer.toString(Integer.MIN_VALUE).toCharArray();
+        private static final char[] MAX_LONG_IMAGE = Long.toString(Long.MAX_VALUE).toCharArray();
+        private static final char[] MIN_LONG_IMAGE = Long.toString(Long.MIN_VALUE).toCharArray();
+        private static final char[] MAX_BINARY_INT_IMAGE = Integer.toBinaryString(Integer.MAX_VALUE).toCharArray();
+        private static final char[] MIN_BINARY_INT_IMAGE = ("-" + Integer.toBinaryString(Integer.MIN_VALUE)).toCharArray();
+        private static final char[] MAX_BINARY_LONG_IMAGE = Long.toBinaryString(Long.MAX_VALUE).toCharArray();
+        private static final char[] MIN_BINARY_LONG_IMAGE = ("-" + Long.toBinaryString(Long.MIN_VALUE)).toCharArray();
+        private static final char[] MAX_HEX_INT_IMAGE = Integer.toHexString(Integer.MAX_VALUE).toCharArray();
+        private static final char[] MIN_HEX_INT_IMAGE = ("-" + Integer.toHexString(Integer.MIN_VALUE)).toCharArray();
+        private static final char[] MAX_HEX_LONG_IMAGE = Long.toHexString(Long.MAX_VALUE).toCharArray();
+        private static final char[] MIN_HEX_LONG_IMAGE = ("-" + Long.toHexString(Long.MIN_VALUE)).toCharArray();
+
+        abstract boolean isInt(String image, int len);
+        abstract boolean isLong(String image, int len);
+
+        private static boolean valueWithinBounds(String value, int len, char[] minImage, char[] maxImage)
+        {
+            boolean negative = value.charAt(0) == '-';
+            char[] boundaryImage = negative ? minImage : maxImage;
+            int maxImageLength = boundaryImage.length;
+            return len < maxImageLength || (len == maxImageLength && magnitudeLessThanOrEqualTo(value, len, boundaryImage));
```

Comment:
This is pretty complicated logic... I'm hoping to find a whole bunch of unit tests!